draft-ietf-oauth-selective-disclosure-jwt.md

```diff
@@ -659,35 +659,62 @@ the claims `given_name`, `family_name`, and `address`, as it would be sent from
+ * all Disclosures are correct, i.e., their digests are referenced in the issuer-signed JWT.
 
-The Holder MUST perform the following (or equivalent) steps when receiving
-a Combined Format for Issuance:
+The Holder and the Verifier perform the following (or equivalent) steps when receiving
+an SD-JWT:
+
+ 1. Separate the SD-JWT into the issuer-signed JWT and the Disclosures (if any).
+ 2. Validate the issuer-signed JWT:
+    1. Ensure that a signing algorithm was used that was deemed secure for the application. Refer to [@RFC8725], Sections 3.1 and 3.2 for details. The `none` algorithm MUST NOT be accepted.
+    2. Validate the signature over the issuer-signed JWT.
+    3. Validate the Issuer and that the signing key belongs to this Issuer.
+    4. Check that the issuer-signed JWT is valid using `nbf`, `iat`, and `exp` claims, if provided in the SD-JWT, and not selectively disclosed.
+    5. Check that the `_sd_alg` claim value is understood and the hash algorithm is deemed secure.
+ 3. Process the Disclosures and embedded digests in the issuser-signed JWT as follows:
+    1. For each Disclosure provided:
+       1. Calculate the digest over the base64url-encoded string as described in (#hashing_disclosures).
+    2. (*) Identify all embedded digests in the issuer-signed JWT as follows:
+       1. Find all objects having an `_sd` key that refers to an array of strings.
+       2. Find all array elements that are objects with one key, that key being `...` and referring to a string.
+    3. (**) For each embedded digest found in the previous step:
+       1. Compare the value with the digests calculated previously and find the matching Disclosure. If no such Disclosure can be found, the digest MUST be ignored.
+       2. If the digest was found in an object's `_sd` key:
+          1. If the respective Disclosure is not a JSON-encoded array of three elements, the SD-JWT MUST be rejected.
+          2. Insert, at the level of the `_sd` key, a new claim using the claim name and claim value from the Disclosure.
+          3. If the claim name already exists at the same level, the SD-JWT MUST be rejected.
+          4. Recursively process the value using the steps described in (*) and (**).
+       3. If the digest was found in an array element:
+          1. If the respective Disclosure is not a JSON-encoded array of two elements, the SD-JWT MUST be rejected.
+          2. Replace the array element with the claim value from the Disclosure.
+          3. Recursively process the value using the steps described in (*) and (**).
+    4. If any digests were found more than once in the previous step, the SD-JWT MUST be rejected.
+    5. Remove all array elements for which the digest was not found in the previous step.
+    6. Remove all `_sd` keys and their contents from the issuer-signed JWT payload.
+    7. Remove the claim `_sd_alg` from the SD-JWT payload.
```

Comment:
These should be applicable only to the Verifier. If we leave these here it sounds as if the Holder is supposed to modify the signed SD-JWT before passing it on to the Verifier.

Reply:
We definitely don't want to suggest that the Holder is supposed to modify the signed SD-JWT before passing it on to the Verifier. But these steps are conceptually applicable to the Holder too so it can access the "processed SD-JWT payload." 

There are also a few steps intertwined above that have "Insert [... ] a new claim" or "Replace the array element" or "Remove all" that would be awkward to pull out into the Verifier only part.   Can we instead somewhere/somehow say that the steps that suggest changing the payload aren't actually touching the SD-JWT payload itself but rather processing the payload content that will be provided to the application?

Reply:
If the Wallet/Holder is trying to display the claims received in the SD-JWT to the user, it needs to remove all the redundant claims IMO, so I think we should keep these steps here, but probably add a sentence saying that `the Holder MUST pass the issuer-signed JWT unmodified to the Verifier`?